Help: Track page views while users are chatting with olark operators.

omarjackman opened the pull request:
This pull request allows Happiness Engineers to see the URL of the page being viewed by the user they are chatting with.

Olark operators will now see something similar to this:

![screen shot 2015-12-17 at 5 56 51 am](https://cloud.githubusercontent.com/assets/1854440/11868046/148b9ea8-a483-11e5-822e-fd98b5f63f15.png)

Fixes #1506

mtias commented on client/lib/olark/index.js:
```diff
@@ -373,3 +377,13 @@ const olark = {
 emitter( olark );
 olark.initialize();
+
+export function trackPageViews( context, next ) {
```
I'd call this `notifyLocation` or something like that instead. Reading it in `boot` it made me think of stats instead.

omarjackman commented:
I found some issues here after doing some more testing. I'll submit for review again after its fixed.

omarjackman commented:
> I'd call this notifyLocation or something like that instead. Reading it in boot it made me think of stats instead.

@mtias I renamed the function and fixed an issue with notifying with the wrong url. Please review again

omarjackman commented:
@mtias @dllh @dmsnell any objections to shipping this?

dmsnell commented on client/lib/olark/index.js:
```diff
@@ -177,6 +178,9 @@ const olark = {
 
+		// system.give_location_to_operator doesn't work well with our single page app so we do
+		// it our selves using page context in the notifyLocation function
+		olarkApi.configure( 'system.give_location_to_operator', false );
```
thank you for this comment - this is a _great_ comment!

dmsnell commented on client/lib/olark/index.js:
```diff
@@ -368,8 +372,28 @@ const olark = {
+
+export function notifyLocation( context, next ) {
+	// Use context.path here because window.location.href has the previous location and not what the new location will be.
+	const location = `${window.location.protocol}//${window.location.host}${context.path}`;
```
In case we could run into issues with other sneaky parts of the URL, I'm pretty sure we could use the `url` library here and only replace the part we want instead of rebuilding from scratch…

```js
import url from 'url';

const location = url.resolve( url.parse( window.location ), context.canonicalPath );
```

This seems to work in my node console and then we wouldn't have to worry about protocol or port or query string or anything like that.

dmsnell commented on client/lib/olark/index.js:
```diff
@@ -368,8 +372,28 @@ const olark = {
+	},
+
+	notifyLocation( location ) {
+		const { details } = olarkStore.get();
```
you can cut out a few lines here and you won't run into exceptions if these don't exist because of the soft-fail modes of destructuring

```js
const { details: { isConversing = false } } = olarkStore.get();

if ( ! isConversing ) {
	return;
}
```

dmsnell commented:
@omarjackman looks like a great feature - I'm sure the HEs will love it! Anyway, check out my note on the URL piece, that seems a bit risky the way it's written (not like it would break anything if a bug appeared though)

omarjackman commented:
@dmsnell I made the suggested changes.

omarjackman commented:
I'll be merging this tomorrow AM if there are no objections by then

dllh commented:
The change seems reasonable to me. :ship:

omarjackman commented:
Had to revert this because of a missing `!` that went missing during my last rebase
